src/NHibernate/Cache/ICacheConcurrencyStrategy.cs

```diff
@@ -175,23 +180,37 @@ public static object[] GetMany(this ICacheConcurrencyStrategy cache, CacheKey[]
 		public static bool[] PutMany(this ICacheConcurrencyStrategy cache, CacheKey[] keys, object[] values, long timestamp,
 		                          object[] versions, IComparer[] versionComparers, bool[] minimalPuts)
 		{
-			if (!(cache is IBatchableCacheConcurrencyStrategy batchableCache))
+			if (cache is IBatchableCacheConcurrencyStrategy batchableCache)
+			{
+				return batchableCache.PutMany(keys, values, timestamp, versions, versionComparers, minimalPuts);
+			}
+
+			var result = new bool[keys.Length];
+			for (var i = 0; i < keys.Length; i++)
 			{
-				throw new InvalidOperationException($"Cache concurrency strategy {cache.GetType()} does not support batching");
+				result[i] = cache.Put(keys[i], values[i], timestamp, versions[i], versionComparers[i], minimalPuts[i]);
 			}
-			return batchableCache.PutMany(keys, values, timestamp, versions, versionComparers, minimalPuts);
+
+			return result;
 		}
 
-		public static bool IsBatchingGetSupported(this ICacheConcurrencyStrategy cache)
+		// 6.0 TODO: remove
+		internal static bool PreferMultipleGet(this ICacheConcurrencyStrategy cache)
 		{
-			// ReSharper disable once SuspiciousTypeConversion.Global
-			return cache.Cache is IBatchableReadOnlyCache && cache is IBatchableCacheConcurrencyStrategy;
+			if (cache is IBatchableCacheConcurrencyStrategy batchableCache)
+				return batchableCache.CacheBase.PreferMultipleGet;
+			return false;
 		}
 
-		public static bool IsBatchingPutSupported(this ICacheConcurrencyStrategy cache)
+		// 6.0 TODO: remove
+		internal static CacheBase GetCacheBase(this ICacheConcurrencyStrategy cache)
 		{
-			// ReSharper disable once SuspiciousTypeConversion.Global
-			return cache.Cache is IBatchableCache && cache is IBatchableCacheConcurrencyStrategy;
+			if (cache is IBatchableCacheConcurrencyStrategy batchableCache)
+				return batchableCache.CacheBase;
+			var concreteCache = cache.Cache;
+			if (concreteCache == null)
+				return null;
+			return concreteCache as CacheBase ?? new ObsoleteCacheWrapper(concreteCache);
```

Comment:
This allows a simplification, by always yielding a `CacheBase` provided there is a cache, even if it does not implement itself `CacheBase`.
In most cases it will yield the wrapper held by the cache concurrency, avoiding the instantiation here. (The instantiation will occur only if the user uses a custom cache concurrency, which looks very unlikely to me, since the mapping does not allow specifying a custom implementation.)